### PR TITLE
Pointer passing examples

### PR DESCRIPTION
**Pull request by trueqbit**
Solidified pointer-passing interface.

- Added example for statically and dynamically bound objects.
- Added unit tests for dynamically allocated objects.
- Fixed UB and refactored obtaining 'xDestroy' function.
  * there's now a safe way of obtaining a 'xDestroy' function from a deleter and a pointer.
  * the previous approach was inherently UB on some platforms (function pointers can't be aliased and then called through that alias, unless the argument/return type representation is the same).
  * implemented understandable constraints which 'deleter' can be used for obtaining a 'xDestroy' function.
  * stateless deleters are possible, as well as objects yielding a 'xDestroy' function pointer.

### CODE REVIEW

**Comment by fnc12** [comment]
@trueqbit when I should make a review? [Now or later](https://www.youtube.com/watch?v=La8_Q6YBTSs&ab_channel=DylanMatthew)?

**Comment by trueqbit** [comment]
> @trueqbit when I should make a review?

Wait still, please. I am still investigating matters in regard to pointer-to-object lifetime!

**Comment by trueqbit** [comment]
@fnc12 F.I.N.A.L.L.Y.
woah. C++20 builds are slow!

**Comment by fnc12** [comment]
@trueqbit so is it time to review or no?

**Comment by trueqbit** [comment]
@fnc12 Yes, it's time for review - sorry I had those ideas overnight, which made it into the last commit.

I am inclined to remove the `template<const char* T> bindable_pointer()`> overloads. It's probably better to change later if needed rather than over-engineer now.

**Comment by fnc12** [comment]
@trueqbit please review my comment at `examples/pointer_passing_interface.cpp` at line 207

**Comment by trueqbit** [comment]
@fnc12 Where's your comment? I presume it's about the commented out order-by clause - I couldn't get it to work initially. Seems like I misunderstood multi_order_by.

Do you have other comments before I push again, and trigger another test build?

In case you didn't see my previous question:
> I am inclined to remove the `template<const char* T> bindable_pointer()`> overloads. It's probably better to change later if needed rather than over-engineer now.

I think you agree, but I'd like to make sure before I remove it...

**Comment by fnc12** [comment]
> I am inclined to remove the template<const char* T> bindable_pointer()> overloads. It's probably better to change later if needed rather than over-engineer now.

Yes I agree.

Yes my comment is about multy_order_by. What result you get if you uncomment it? Let's fix it or remove it. Commented code in examples is not the best idea.

Please push new code at any time. Thank you for your job. It is awesome.

**Comment by trueqbit** [comment]
The wrong sort order is unrelated to this PR or the pointer-passing feature. I will come up with another bug ticket in regard to this.
Otherwise the PR is good to go once the new commits yield a green test build.

**Comment by fnc12** [comment]
@trueqbit ok. Please review comment at line 698 in `tests/tests2.cpp`

**Comment by fnc12** [comment]
> I reckon it would be helpful if you commented directly at the line in the 'files' view

This is what I do. Thanks

**Comment by fnc12** [comment]
Снимок экрана 2021-11-24 в 21 02 16

**Comment by trueqbit** [comment]
> > I reckon it would be helpful if you commented directly at the line in the 'files' view
> 
> This is what I do. Thanks

How dare I to ask you :) I swear that I didn't see anything; also now I don't see any comments or conversation :(

**Comment by fnc12** [comment]
Strange. Probably I must make a review finish for you to see my comments

**Review by fnc12 — CHANGES_REQUESTED** [review]

**Comment by fnc12** [comment]
check it out now please

File: examples/pointer_passing_interface.cpp [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+                    as<str_alias<'o', 'k'>>(c(&Result::errorValue) == 0),
+                    &Result::errorValue,
+                    &Result::errorCategory,
+                    as<str_alias<'e', 'q'>>(func<equal_error_code_fn>(
+                        func<make_error_code_fn>(&Result::errorValue, &Result::errorCategory),
+                        bindable_pointer<ecode_pvt>(new error_code{}, default_delete<error_code>{}))),
+                    func<error_category_name_fn>(func<get_error_category_fn>(&Result::errorCategory)),
+                    func<error_category_message_fn>(func<get_error_category_fn>(&Result::errorCategory),
+                                                    &Result::errorValue))/*,
+            multi_order_by(!c(&Result::errorValue), &Result::errorCategory, &Result::errorValue, &Result::id)*/);
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply examples/pointer_passing_interface.cpp]: Corrected in FireDaemon/sqlite_orm@a2bb4fcfc48b.

No, it's not a valid expression. I believe it should - it is valid sql and makes sense when columns are falsy [null or 0]. But it requires expression_t to be negatable.

File: tests/tests2.cpp [hunk 2]
```diff
@@ -601,19 +651,196 @@ TEST_CASE("Pointer") {
...
+                REQUIRE(delete_int64::deleted == false);
+                storage.execute(stmt);
+                REQUIRE(delete_int64::deleted == false);
+            }
+            // bound pointer value must be deleted when prepared statement is going out of scope
+            REQUIRE(delete_int64::deleted == true);
+        }
+
+        SECTION("ownership transfer") {
+            {
```
[the remark this answers is not in the document]
Comment (trueqbit) [reply tests/tests2.cpp]: I thought I leave the brackets for the sake of resemblance to the previous sections.
No objection, though :)
Corrected in commit FireDaemon/sqlite_orm@2547bfe.

**Comment by fnc12** [comment]
please add `tests/statement_serializator_tests/bound_pointer.cpp` to CMakeLists.txt

**Review by fnc12 — APPROVED** [review]

**Comment by fnc12** [comment]
@trueqbit thanks! This is a great feature!

**Pull request merged**